src/processor/tree/mod.rs

```diff
@@ -0,0 +1,73 @@
+mod snapshot;
+mod tree_wrapper;
+
+use std::path::Path;
+
+use eyre::Result;
+use tokio::{sync::mpsc, task::JoinHandle};
+
+use crate::{constants::ethereum::STATE_FILE_PATH, types::CommitBlockInfoV1};
+
+use self::{snapshot::StateSnapshot, tree_wrapper::TreeWrapper};
+
+use super::Processor;
+
+pub struct TreeProcessor<'a> {
+    tree: TreeWrapper<'a>,
+    snapshot: StateSnapshot,
+}
+
+impl TreeProcessor<'static> {
+    pub fn new(db_dir: &Path) -> Result<Self> {
+        // TODO: Implement graceful shutdown.
+        // If database directory already exists, we try to restore the latest state.
+        // The state contains the last processed block and a mapping of index to key
+        // values, if a state file does not exist, we simply use the defaults instead.
+        let should_restore_state = db_dir.exists();
+        let snapshot = if should_restore_state {
+            println!("Loading previous state file...");
+            StateSnapshot::read(STATE_FILE_PATH).expect("state file is malformed")
+        } else {
+            println!("No existing database found, starting from genesis...");
+            StateSnapshot::default()
+        };
+
+        // Extract `index_to_key_map` from state snapshot.
+        let StateSnapshot {
+            ref index_to_key_map,
+            ..
+        } = snapshot;
+
+        let tree = TreeWrapper::new(db_dir, index_to_key_map.clone())?;
+
+        Ok(Self { tree, snapshot })
+    }
+}
+
+impl Processor for TreeProcessor<'static> {
+    fn run(mut self, mut rx: mpsc::Receiver<Vec<CommitBlockInfoV1>>) -> JoinHandle<()> {
+        tokio::spawn(async move {
```

Comment:
I'm not experienced in Rust nor async Rust, but would this structure be slightly more flexible & easier to use if this was `async fn` and the task spawning was done on higher level? It would also make the trait interface more ubiquitous in the sense that it wouldn't involve async / Tokio specifics there, I think?